src/main/java/ch/heigvd/pro/b04/sessions/SessionController.java

```diff
@@ -128,18 +133,59 @@ public ServerSession putSession(
       throw new PollNotExistingException();
     }
 
-    // 3. Get the session or create one
     ServerPoll poll = pollList.get(0);
-    List<ServerSession> sessionList = sessionRepository.findByModAndPoll(modFromId.get(), poll);
 
-    ServerSession session = sessionList.isEmpty() ? poll.newSession(sessionRepository)
-                                                  : sessionList.get(0);
-
-    // 4. Update the session
-    session.setState(clientSession.getState());
-
-    sessionRepository.saveAndFlush(session);
+    Optional<ServerSession> last = poll.getLatestSession(sessionRepository);
+    ServerSession session;
+
+    SessionState askedState = clientSession.getState();
+
+    if (last.isEmpty()) {
+      if (! askedState.equals(SessionState.OPEN)) {
+        throw new SessionStateMustBeOpenedFirstException();
+      } else {
+        session = poll.newSession(sessionRepository);
+      }
+    } else {
+      session = last.get();
+      if (session.getState().equals(SessionState.OPEN)) {
+        switch (askedState) {
+          case OPEN:
+            throw new SessionStateMustBeClosedFirstException();
+          case QUARANTINED:
+            session.setState(askedState);
+            break;
+          case CLOSED:
+            session.close();
+            break;
+          default:
+            throw new IllegalSessionStateException();
+        }
+      } else if (session.getState().equals(SessionState.QUARANTINED)) {
+        switch (askedState) {
+          case OPEN:
+          case QUARANTINED:
+            throw new SessionStateMustBeClosedFirstException();
```

Comment:
We probably have to discuss whether this is the behavior we expect in class. Don't we want to offer flexibility to the user and let them open the poll "again" with the same access code ?